### ISSUE TITLE
cupy.asarray slower with pinned memory

### ISSUE THREAD

**Issue by RapidEdgeAI** [cupy/cupy#7532]
### Description

Hi,

tl;dr:
when I create a numpy array on pinned memory and copy it to the GPU using cupy.asarray or cupy.array it allocates new memory on the host and copies the numpy array which makes using pinned memory slower than not using it.

long version:
When you create a normal numpy array (not pinned memory) and copy it to the GPU using cupy.asarray or cp.array it creates pinned memory/page locked memory first, copies the array to the pinned memory and then copies it into the VRAM of the GPU (Pageable Data Transfer) . 
![image](https://user-images.githubusercontent.com/111732409/233855314-88c6440d-1e2b-4e1b-b5d3-25a73cf05b8d.png)
(source: https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/)
If you do the same thing again cupy.asarray skips the allocation step because it already has been done. This is the behaviour which i expected. I observed this behaviour in NVIDIA Nsight Systems 2023.2.1.
But there is a problem if you use a numpy array which is located on pinned memory (Pinned Data Transfer). If you want to copy it with cupy.asarray it doesn't recognize it as pinned memory an allocates new memory on the host. This allocation takes relatively long time making the my code slower when using pinned memory.
This allocation does not happen when using cupy.asnumpy with a numpy array on pinned memory for the "out" parameter. Because of this the mentioned problem i think cupy.asarray and cupy.array have a bug.

![image](https://user-images.githubusercontent.com/111732409/233855599-14e97163-1b75-42e9-b790-20b60ad67559.png)

To avoid the unnecessary allocation of new pinned memory I am using img_gpu = cupy.empty_like(img) to get memory from the gpu memory pool and then img_gpu.set(img) to copy the data from the pinned host memory to the GPU.

Im using all cupy functions within a non-blocking stream to avoid synchronization after each cupy call.
I am using cupy 12.0.0.

Please see my commented test code for further information.

### To Reproduce

```py
import cupy as cp
import cupyx as cpx
import numpy as np
import time

print(cp.__version__)

def compute(img, stream, out):

    with stream:
        # img_gpu = cp.asarray(img) # copy img to gpu
        
        # cupy.asarray is the recommended way to copy data from the host to the gpu. However, it seems to have a bug
        # in the current cupy version (12.0.0). It doesn't recognize pinned memory. Instead, it allocates new
        # pinned memory and copies the data. This causes great overhead and makes working with pinned memory even
        # slower than working without pinned memory. This issue does not happen when using cupy.asnumpy to copy
        # from gpu to pinned memory on the host.
        # Because of the supposed bug, the following workaround is used which avoids
        # the overhead:
        img_gpu = cp.empty_like(img) # get memory from cupy's gpu memory pool
        img_gpu.set(img) # copy img to gpu

        # do some async operations to fill the queue of the stream
        img_gpu = img_gpu.astype(cp.float32) # change dtype
        img_gpu /= 314.2
        img_gpu *= 2
        img_gpu = img_gpu.transpose((2, 0, 1))[::-1]  # HWC to CHW, BGR to RGB

        img_cpu_new = cp.asnumpy(img_gpu, out=out) # copy result back to cpu

    stream.synchronize() # wait for the completion of queued work

    return img_cpu_new

def print_time(start,end, text):
    print(text, (end-start), "s")

def do_test(img, text, stream, out=None):
    start = time.time()
    img_result = compute(img, stream, out)
    end = time.time()
    print_time(start, end, text)

stream = cp.cuda.Stream() # cuda stream
shape = (1920, 1080, 3)
shape_result = (shape[2], shape[0], shape[1])

# warmup
img_cpu = np.random.randint(low=0, high=255, size=shape)  # image
do_test(img_cpu, "warmup:", stream)


# without pinned memory
img_cpu = np.random.randint(low=0, high=255, size=shape)  # image
do_test(img_cpu, "without pinned memory:", stream)

# with pinned memory
img_cpu = np.random.randint(low=0, high=255, size=shape)  # image
img_cpu_pinned = cpx.empty_like_pinned(img_cpu) # allocate pinned memory (aka page-locked memory)
img_cpu_pinned[:] = img_cpu # copy the image into the pinned memory
out_cpu_pinned = cpx.empty_pinned(shape_result, dtype=cp.float32) # allocate pinned memory for the result
do_test(img_cpu_pinned, "with pinned memory:", stream, out_cpu_pinned)

# profile this script using: 
# nsys profile /usr/bin/python3 path_to_this_script/test_streams.py
# then open the generated *.nsys-rep file in NVIDIA Nsight Systems GUI
```

### Installation

Wheel (`pip install cupy-***`)

### Environment

```
# Paste the output here
```

**Comment by takagi** [comment]
Thanks for raising this issue. Currently, we have `cupy.asarray` and `cupy.array` work as such to avoid potential data race as discussed:
- https://github.com/cupy/cupy/blob/6668d290b33a65e18fecf8b817e2af839f31b7b3/cupy/_core/core.pyx#L2545-L2548
- https://github.com/cupy/cupy/pull/5155#discussion_r621808782

BTW, CuPy also have the memory pool for pinned memory, so the staging memory should be reused even when you explicitly allocate pinned memory on the host. Doesn't it work for you?

**Comment by RapidEdgeAI** [comment]
Hi @takagi,
thank you very much for your response this is immensely helpful. I'm sorry for thinking that this was a bug.

I am not using the memory pool because it wanted to avoid coping images from paged memory to pinned / page-locked-memory on the CPU each time i upload an image to the GPU to reduce CPU load. Instead I wanted to allocate my images directly on pinned memory. In my actual code for image processing, I am reading video frames directly into the pinned memory using OpenCV.
```
import cv2
import cupy
import cupyx

cap = cv2.VideoCapture("path_to_video/video.avi")
pinned_mem = cupyx.empty_pinned((1920, 1080, 3), dtype=cupy.uint8) # allocate pinned memory
ret, frame = cap.read(pinned_mem) # read frame directly into pinned memory
frame_gpu = cupy.asarray(frame) # img from cpu to gpu

# some computations using cupy ...
```

**Comment by kmaehashi** [comment]
CuPy uses the memory pool for pinned memory by default. The second allocation of `cupyx.empty_pinned()` should be fast if the memory chunk is reclaimed from the pool.

**Comment by RapidEdgeAI** [comment]
Hello @kmaehashi,
thank you! This cleared up my confusion.
